_posts/2022-12-12-Remove-Unused-Code-to-Avoid-Decay.md

```diff
@@ -0,0 +1,126 @@
+
+<i class="fas fa-hiking" style="font-size:4em;"></i>
+{:.center}
+
+Think about your project as if it were on a hiking trail: __"Carry as little as possible, but choose that little with care."__ (Earl Shaffer). Be careful what code you choose to carry; it may impact your ability to address needs over time and lead to otherwise unintended software decay.
+
+## Detecting Unused Code with Vulture
+
+Understanding the cost of added content, it's important to routinely examine which parts of your code are still necessary. You can prepare your code for a long journey by detecting (and removing) unused code with various automated tools. These tools are generally designed for static analysis and linting, meaning they may also be incorporated into automated and routine testing.
+
+```shell
+$ vulture unused_code_example.py
+unused_code_example.py:3: unused import 'os' (90% confidence)
+unused_code_example.py:4: unused import 'pd' (90% confidence)
+unused_code_example.py:7: unused function 'unused_function' (60% confidence)
+unused_code_example.py:14: unused variable 'unused_var' (60% confidence)
+```
+
+_Example of Vulture command line usage to discover unused code._
+
+[Vulture](https://github.com/jendrikseipp/vulture) is one tool dedicated to finding unused python code. Vulture provides both a command line interface and Python API for discovering unused code. It also provide a rough confidence to show how certain it was about whether the block of code was unused. See the following interactive example for a demonstration of using Vulture.
+
+[Example Link](https://cu-dbmi.github.io/notebooks/lab?path=unused_code_detection.ipynb)
+
+## Further Code Usefulness Detection with Pylint and Coverage.py
+
+In addition to Vulture, [Pylint](https://pylint.pycqa.org/en/latest/index.html) and [Coverage.py](https://coverage.readthedocs.io/) can be used in a similar way to help show where code may not have been used within your project.
+
+[Pylint](https://pylint.pycqa.org/en/latest/index.html) focuses on code style and other static analysis in addition to unused variables. See [Pylint's Checkers](https://pylint.pycqa.org/en/latest/user_guide/checkers/features.html) page for more details here, using "unused-*" as a reference to checks it performs which focus on unused code.
+
+[Coverage.py](https://coverage.readthedocs.io/) helps show you which parts of your code have been executed or not. A common usecase for Coverage involves measuring "test coverage", or which parts of your code are executed in relationship to tests written for that code. This provides another perspective on code utility; if there's not a test for the code, is it worth keeping?
```

Comment:
Extremely minor nitpick, but I prefer a colon over a semicolon if the clause following it is heavily implied, i.e.:
```suggestion
[Coverage.py](https://coverage.readthedocs.io/) helps show you which parts of your code have been executed or not. A common usecase for Coverage involves measuring "test coverage", or which parts of your code are executed in relationship to tests written for that code. This provides another perspective on code utility: if there's not a test for the code, is it worth keeping?
```

Reply:
Thank you, this makes sense to me and feels like it flows better. I'll add this in.